analytic_engine/src/sst/parquet/encoding.rs

```diff
@@ -461,3 +487,457 @@ impl ParquetEncoder {
+            }
+            bitmap_length_so_far += value_num as usize;
+            new_values
+                .extend(value_slices[*prev as usize..*current as usize].repeat(value_num as usize));
+            for _ in 0..value_num {
+                value_length_so_far += value_len;
+                new_offsets.push(value_length_so_far);
+            }
+        }
+        debug!(
+            "new buffer slice, offsets:{:#02x?}, values:{:#02x?}, bitmap:{:#02x?}",
+            new_offsets.as_slice(),
+            new_values.as_slice(),
+            new_null_buffer.as_slice(),
+        );
+
+        let array_data = ArrayData::builder(array_ref.data_type().clone())
+            .len(values_num)
+            .add_buffer(new_offsets.into())
+            .add_buffer(new_values.into())
+            .null_bit_buffer(Some(new_null_buffer.into()))
+            .build()
+            .map_err(|e| Box::new(e) as _)
+            .context(DecodeRecordBatch)?;
+
+        Ok(array_data.into())
+    }
+
+    /// Like `stretch_variable_length_column`, but array value is fixed-size
+    /// type.
+    fn stretch_fixed_length_column(
+        array_ref: &ArrayRef,
+        value_size: usize,
+        value_offsets: &[i32],
+    ) -> Result<ArrayRef> {
+        assert!(!value_offsets.is_empty());
+
+        let values_num = *value_offsets.last().unwrap() as usize;
+        let raw_buffer = array_ref.data().buffers()[0].as_slice();
+        let null_bitmap = array_ref.data().null_bitmap();
+
+        let mut new_buffer = MutableBuffer::new(value_size * values_num);
+        let mut new_null_buffer = MutableBuffer::new(values_num);
+        new_null_buffer = new_null_buffer.with_bitset(values_num, true);
+        let null_slice = new_null_buffer.as_slice_mut();
+
+        let mut length_so_far = 0;
+        for (idx, offset) in (0..raw_buffer.len()).step_by(value_size).enumerate() {
+            let value_num = value_offsets[idx + 1] - value_offsets[idx];
+            if let Some(bitmap) = null_bitmap {
+                if !bitmap.is_set(idx) {
+                    for i in 0..value_num {
+                        bit_util::unset_bit(null_slice, length_so_far + i as usize);
+                    }
+                }
+            }
+            length_so_far += value_num as usize;
+            new_buffer.extend(raw_buffer[offset..offset + value_size].repeat(value_num as usize))
+        }
+        let array_data = ArrayData::builder(array_ref.data_type().clone())
+            .add_buffer(new_buffer.into())
+            .null_bit_buffer(Some(new_null_buffer.into()))
+            .len(values_num)
+            .build()
+            .map_err(|e| Box::new(e) as _)
+            .context(DecodeRecordBatch)?;
+
+        Ok(array_data.into())
+    }
+
+    /// Returns offsets of nested items in ListArray, and populates the
+    /// `values_num`
+    ///
+    /// Note: caller should ensure `array_ref` is ListArray
+    fn get_list_array_offsets(array_ref: &ArrayRef) -> Vec<i32> {
```

Comment:
Offset in arrow is a generic parameter. It can be `i32` or `i64`. The constant `OFFSET_SIZE` should also consider this